src/main/java/htsjdk/variant/variantcontext/JEXLMap.java

```diff
@@ -11,152 +12,188 @@
      * @return a collection of boolean values, representing the results of all the variants evaluated
+     *
+     * @throws IllegalArgumentException
      */
     public Collection<Boolean> values() {
-        // this is an expensive call
-        for (JexlVCMatchExp exp : jexl.keySet())
-            if (jexl.get(exp) == null)
+        for (final JexlVCMatchExp exp : jexl.keySet()) {
+            if (jexl.get(exp) == null) {
                 evaluateExpression(exp);
+            }
+        }
         return jexl.values();
     }
 
     /**
-     * evaulate a JexlVCMatchExp's expression, given the current context (and setup the context if it's null)
-     * @param exp the JexlVCMatchExp to evaluate
+     * TODO: the number may count invalid entries, is this good?
+     * @return the number of keys, i.e. {@link JexlVCMatchExp}'s hold by this mapping.
+     */
+    public int size() {
+        return jexl.size();
+    }
+
+    public boolean isEmpty() { return this.jexl.isEmpty(); }
+
+    // -----------------------------------------------------------------------------------------------
+    // Modifiers
+    // -----------------------------------------------------------------------------------------------
+
+    public Boolean put(JexlVCMatchExp jexlVCMatchExp, Boolean aBoolean) {
+        return jexl.put(jexlVCMatchExp, aBoolean);
+    }
+
+    public void putAll(Map<? extends JexlVCMatchExp, ? extends Boolean> map) {
+        jexl.putAll(map);
+    }
+
+    // -----------------------------------------------------------------------------------------------
+    // Utilities
+    // -----------------------------------------------------------------------------------------------
+
+    /**
+     * Lazily initializes {@link #jexl}, in the sense that all keys in {@code jexlCollection} are associated with {@code null}.
+     * @param jexlCollection
+     */
+    private void lazyInitialize(Collection<JexlVCMatchExp> jexlCollection) {
+        jexl = new HashMap<>();
+        for (final JexlVCMatchExp exp: jexlCollection) {
+            jexl.put(exp, null);
+        }
+    }
+
+    /**
+     * Evaluates a {@link JexlVCMatchExp}'s expression, given the current context (and setup the context if it's {@code null}).
+     * @param exp the {@link JexlVCMatchExp} to evaluate
+     * @throws IllegalArgumentException when {@code exp} is {@code null}
      */
-    private void evaluateExpression(JexlVCMatchExp exp) {
+    private void evaluateExpression(final JexlVCMatchExp exp) {
         // if the context is null, we need to create it to evaluate the JEXL expression
-        if (this.jContext == null) createContext();
+        if (this.jContext == null){
+            createContext();
+        }
+
         try {
             final Boolean value = (Boolean) exp.exp.evaluate(jContext);
             // treat errors as no match
             jexl.put(exp, value == null ? false : value);
-        } catch (Exception e) {
+        } catch (final JexlException e) {
```

Comment:
In the previous version this handled exceptions that had "undefined variable" differently than other exceptions.  I think you should have a separate catch block for `JexlException.Variable` that does `jexl.put(exp,false)` and then the catch for generic `JexlException` that get translated to `IllegalArgumentException`